### DISCUSSION TITLE
[Proposal] atomWithWrapper

### DISCUSSION THREAD

**Discussion by TwistedMinda** [pmndrs/jotai#1172] [Ideas] [answered]
Hello,

I'm currently trying to setup a clean opinionated approach for using jotai, cause I'm finding approximately none in the wild. Planning to make the big step & start using it in production ASAP.
My constraints are the following:
- separate groups of logic in "modules"
- keep component/modules imports untied to underlying library
- typescript must be respected all the way up

I've come up with 2 main elements:

- A "State" wrapper for my exposed atoms :
```tsx
type Writable<Value> = WritableAtom<Value, any, any>
type Store<Value, FamilyParam = any> = Atom<Value> | Writable<Value> | AtomFamily<FamilyParam, Atom<Value>>

export class State<Value, FamilyParam> {
  reference: Store<Value>

  constructor(reference: Store<Value, FamilyParam>) {
    this.reference = reference
  }

  ref(param?: any) {
    if (typeof this.reference === 'function')
      return (this.reference as AtomFamily<FamilyParam, Atom<Value>>)(param) // It was an atom family
    return this.reference
  }
  use(param?: any) { return useAtom(this.ref(param) as Atom<Value>) }
  get(param?: any): Value { return useAtomValue(this.ref(param) as Atom<Value>) as Value }
  setter(param?: any) { return useSetAtom(this.ref(param) as Writable<Value>) }
}
// Expose a creator
export function atomWithWrapper<Value, FamilyParam = any>(store: Store<Value, FamilyParam>) {
  return new State<Value, FamilyParam>(store)
}
```
- A "Factory" utility to create atoms :
```tsx
type StateWrite = (get: Getter, set: Setter, payload: any) => any

export class Factory {
  static persistedStore<Value>(key: string, initialValue: Value) {
    return atomWithStorage<Value>(key, initialValue, storage)
  }
  static store<Value>(read: ((get: Getter) => void) | Value, write?: SetStateAction<Value>) {
    if (!write)
      return atom(read) as Atom<Value>
    return atom(read, write) as WritableAtom<Value, SetStateAction<Value>, any>
  }
  static action<Value>(write: StateWrite) {
    return atom(null, write)
  }
  static family<Param, Value>(
    creator: (param: Param) => Atom<Value>,
    compare?: (a: Param, b: Param) => boolean) {
    return atomFamily<Param, Atom<Value>>(creator, compare)
  }
  static wrapped<Value, FamilyParam = any>(store: Store<Value, FamilyParam>) {
    return new State<Value, FamilyParam>(store)
  }
}
```
We can now produce clean logic in a "module" without all the imports :
```typescript
import { Factory, State } from './state'

const count = Factory.persistedStore<number>('persist-count', 0)
const increase = Factory.action(async (get, set) => {
  set(count, prev => prev + 1)
})
const countCreator = Factory.family<number, string>(val => Factory.store(val.toString()))
const wrappedCount = Factory.wrapped(count) // or "atomWithWrapper(count)"

// Expose to the real world
export const CounterModule = {
  count: Factory.wrapped(count),
  countCreator: Factory.wrapped(countCreator),
  wrappedCount, // already wrapped,

  increase: new State(increase), // If you don't like the "Factory.wrapped" nor the "atomWithWrapper"
}
```
And finally the usage in components : 
```tsx
import { CounterModule } from "modules"

const Component = () => {
  const count = CounterModule.count.get()
  const press = CounterModule.increase.setter()
  const memoizedCopy = CounterModule.countCreator.get(count)
  const wrapped = CounterModule.wrappedCount.get()
  return (
    <View>
      <Text>Counter: {count}</Text>
      <Button
        text="Increase me"
        onPress={press}
        />
    </View>
  )
}
```

Contraints respected. I'm pretty happy with the minimalism of the API. But what are you thoughts about this pattern? Am I not thinking jotai-ish enough?

Anyway, I found that it could actually be an util `atomWithWrapper` that would wrap your atom in a class to avoid imports, the atom carries his getter, setter & listener all by itself.
I'm still struggling hard with the typescript tho. I don't know how to easily wrap the call of `atom` without recopying all the same parameters (could be huge and change over time)

What you guys think?

EDIT: Forgot about the `atomFamily` stuff, maybe it doesn't fit, gonna check
EDIT2: Added `atomFamily` stuff, still okay. But broke the typing. `AtomFamily` is not exported from utils grrr
EDIT3: Fixed typings after adding `atomFamily`
EDIT4: Added example with `atomWithWrapper`
EDIT5: Total rework after dai-shi advices

Module logic:
```tsx
import { persist } from './utilis'
import {
  atom, useAtom,
  useAtomValue, useSetAtom,
} from "jotai"

const someUtil = persist('persist-key', 0)
const count = atom(0)

export const CounterModule = {
  useCount: () => useAtomValue(count),
  useSetCount: () => useSetAtom(count),
}
```
Usage:
```tsx
import { CounterModule } from "modules"

const Component = () => {
  const count = CounterModule.useCount()
  const setCount = CounterModule.useSetCount()
  const press = () => {
    setCount(prev => prev + 1)
  }
  return (
    <View>
      <Text>{count}</Text>
      <Button
        text="Increase me"
        onPress={press}
        />
    </View>
  )
}
```

**Comment by dai-shi** [comment, marked as answer]
I think it's just a preference, but I would do something like this.
```ts
type SetStateAction<Value> = Value | ((prev: Value) => Value)

type State<Value> = {
  useValue: () => Value;
  useSetValue: () => (action: SetStateAction<Value>) => void;
};

const createState = <Value>(initialValue: Value): State<Value> => {
  const a = atom(initialValue);
  return {
    useValue: () => useAtomValue(a),
    useSetValue: () => useSetAtom(a),
  };
};
```

It doesn't fulfill your constrains though.
btw, I wouldn't overuse atomFamily, because unless you use it with a finite number of params, there's a memory leak risk.

**Reply by TwistedMinda** [reply]
@dai-shi thanks for your response

Yes I could have used a function instead of class, probably more in the mind of jotai. But it would require to use a creator + the type definition instead of just 1 class definition, right?

I still need to test more atomFamily before I use it in production indeed.

Do you see anything that doesn't fit the philosophy of jotai ?
Maybe you already used it in some other way already ?
Did you export/import your atoms seperately?

Did you ever use it alone without any other lib before? did it scale?

**Reply by dai-shi** [reply]
> Do you see anything that doesn't fit the philosophy of jotai ?

Hm, this is a good question. In my mental model, I would just use ES modules and don't hide atoms.
Because if you look at the atom object (or types), it's already isolated from the implementation. It's just a definition, a set of functions.
So, I would say hiding atoms already doesn't fit the philosophy of jotai.
At the same time, it's unopinionated how to organize your app, so it's fine.

> Maybe you already used it in some other way already ?

I don't for the reason. I know some people hide atoms with React custom hooks.

> Did you export/import your atoms seperately?

Yes, that's what I prefer (if I understand your question correctly.)

> Did you ever use it alone without any other lib before? did it scale?

If I create a jotai app, I would try to make use of jotai as much as possible.
I'm not super experienced with large scale apps. So, there must be something missing. That's why I say the community is welcome to work on jotai best practices.

**Reply by TwistedMinda** [reply]
Alright, I feel you!
Thanks for those details!

I'm still in love with my setup tho.
- The fact I can switch to recoil or a new comer in a day if needed/wanted... I love encapsulation.
- The fact my imports are untied to any library and one-per-module

I think I'm gonna stick to it until more opinions coming up here or in the wild.

**Reply by TwistedMinda** [reply]
@dai-shi I have very much changed my mind about my setup.

The CONS i felt of using my encapsulated setup:
- It feels more duplicating then encapsulating
- I'm wasting time on the typescript typings (even though it teaches a lot)
- Some API like atomFamily seem risky to encapsulate

The PROS of using the library directly
- The API being so minimal already makes it very easy to use
- The API will be mostly used in the "logic" files anyway, not much in the components themselves
- Typescript is of course by default with no work required
- You can still batch the atoms on export as I mention in within "modules"

My constraints are better suited without encapsulation :p I take my words back, not in love of my setup anymore

BUT ! I still feel like it's bad to be forced to always import the same 3 : `useAtom`, `useSetAtom`, `useAtomValue`. As we've discussed it, it would be very easy to add them to the atom skeleton itself, as its 100% gonna be used anyway. I'm gonna go for it on my side.

Do you think it would be a bad idea to do it at the core of the lib ? Just because of the duplication maybe it feels odd ? Very practical tho

**Reply by dai-shi** [reply]
> it's bad to be forced to always import the same 3 : `useAtom`, `useSetAtom`, `useAtomValue`.

I think force importing is fine. It's typical react hooks usage.
I would actually prefer it. It's good for code minification.

```js
const foo = fooObject.useValue();
```
That ☝️ is not nice, because there's no guarantee that fooObject will not be mutated. It can violate the rules of hooks.

> As we've discussed it, it would be very easy to add them to the atom skeleton itself, as its 100% gonna be used anyway. I'm gonna go for it on my side.

So, you mean like this?
```js
const createAtomWithHooks = (read, write) => {
  const anAtom = atom(read, write);
  anAtom.use = () => useAtom(anAtom);
  anAtom.useValue = () => useAtomValue(anAtom);
  anAtom.useSet = () => useSetAtom(anAtom);
  return Object.freeze(anAtom); // by freezing, we can avoid mutation.
};
```
I think it's fine to do it on your end, but...

> Do you think it would be a bad idea to do it at the core of the lib ? 

Yes, because atoms are independent from React.
I have even an experimental project to make jotai work without React.
https://github.com/jotai-labs/jotai-jsx

I will consider making jotai framework-agnostic and separate react dependent entry point in the far future. Like this:
```js
import { atom } from 'jotai/vanilla';
import { useAtom } from 'jotai/react';
```
(Note again that this will not happen anytime soon.)

**Reply by TwistedMinda** [reply]
@dai-shi  This was a very interesting read. Thanks for those information!
I decided to enter your world, and it's still respecting my initial constraints. Why did I over-engineer this so bad, jotai is simple!

Module logic:
```tsx
import { persist } from './utilis'
import {
  atom, useAtom,
  useAtomValue, useSetAtom,
} from "jotai"

const someUtil = persist('persist-key', 0)
const count = atom(0)
const derivedAtom = atom(get => get(count))
const fetch = atom(null, async (get, set, payload) {
  // my action
})

export const CounterModule = {
  useCount: () => useAtomValue(count),
  useSetCount: () => useSetAtom(count),

  get derived() { return useAtomValue(derivedAtom) },
  get fetch() { return useSetAtom(fetch) },
}
```
Usage:
```tsx
import { CounterModule } from "modules"

const Component = () => {
  const count = CounterModule.useCount()
  const setCount = CounterModule.useSetCount()
  const fetch = CounterModule.fetch // I find the "get" variant great for the write-only atoms actually
  const { fetch, derived } = CounterModule // Swag 
  const press = () => {
    fetch()
    setCount(prev => prev + 1)
  }
  return (
    <View>
      <Text>{count}</Text>
      <Button
        text="Increase me"
        onPress={press}
        />
    </View>
  )
}
```

I'm pretty damn happy with what it looks like and how easy and encapsulated it still is. Glad I had this conversation with you before going heavy on useless wraps.

Gonna mark this as resolved as I really think this is the way to go, you got me convinced.

**Reply by dai-shi** [reply]
Cool. Let me add an alternative idea.

```js
// CounterModule.js

export const useCount = () => useAtomValue(count);
export const useSetCount = () => useSetAtom(count);

// Modules.js

export * as CounterModule from './CounterModule';

// Usage

import { CounterModule } from "./Modules"

const Component = () => {
  const count = CounterModule.useCount()
  const setCount = CounterModule.useSetCount()
```

The benefit of this namespace usage is, if your bundler supports ESM, it can minify your code better.

---

```js
  get derived() { return useAtomValue(derivedAtom) },
```
doesn't look good as people can easily misuse it and violate the rules of hooks.

**Reply by TwistedMinda** [reply]
@dai-shi About the ESM optimization. Will that be noticeable if I use under a zillion modules ? :p 
I suppose readability should be the biggest concern here. Need to meditate on what scales best for maintenance:
- a big forest of atoms/selectors at the end of each module (my tactic)
- exporting atoms/selectors as you go (your tactic)

> doesn't look good as people can easily misuse it and violate the rules of hooks.

Indeed! After re-reading myself, it feels odd indeed to treat write atoms differently and provide them with a different convention, it can only lead to more misunderstanding, and potentially crashes if someone would try to refactor naively:
```tsx
const fetchUser = UserModule.fetchUser
const press = () => {
   fetchUser()
}
// into
const press = () => {
  UserModule.fetchUser()
}
```

Which will cause a massive crash. So let's forget about the "get" keyword for atoms. However it can be used if the module exports a normal function, that can help differentiating them from the atoms:
```tsx
export const CounterModule = {
  useCount: () => useAtomValue(count),
  useSetCount: () => useSetAtom(count),
  useDerived: () => useAtomValue(derivedAtom),
  useFetch: () => useSetAtom(fetch),
 
  // Function that doesn't need to get/set the store
  get localStuff() {
     // i'm not sure what could be a good example tho
  }
}
```
Thanks for those details man, very interesting

**Reply by dai-shi** [reply]
I don't disagree readability can be more important than bundle size. My perspective is from a library author, which is a little different from app developers. So, it's fine.

I don't think ESM exports are harder to read than exported objects, but it can be just preference.


> You mean that it's not clear for the developer that they MUST be used inside a component (hence the usual "use" prefix) ? 

And it can't be used conditionally.
```js
const x = someCondition && CounterModule.derived; // violating rules of hooks
```

(edit) you edited, so context lost. 😄

**Reply by TwistedMinda** [reply]
@dai-shi

> My perspective is from a library author, which is a little different from app developers

Yep, totally! You're fulfilling your role perfectly! Just wanted to relativize about the "boost" :p 

> (edit) you edited, so context lost. 😄 

haha yes sorry.
About your example, just to be sure I understood correclty, you mean that you may want to call functions conditonally so you need their reference not just their return value ?
Like, briefly:
```tsx
// Anywhere in code
const x = someCondition ? CounterModule.useDerived : CounterModule.useNormal

const Component = () => {
  // Inside a component
  const value = x()
}
```

Anyway, feels really clear to me that using "get" keyword on atoms was a bad idea, now :p 
Thanks again!

**Reply by dai-shi** [reply]
I meant it's inside render in my example.
```jsx
const Component = () => {
  const x = someCondition && CounterModule.derived; // violating rules of hooks
  return <>{x}</>;
};
```

**Reply by TwistedMinda** [reply]
@dai-shi Sorry but would it not also violate rules of hooks with
```tsx
const Component = () => {
  const x = someCondition && CounterModule.useDerived(); // Error: "React rendered fewer hooks than expected" when "someCondition" changes
  return <>{x}</>;
};
```

**Reply by dai-shi** [reply]
Yes, but that can be captured by eslint-plugin-react-hooks. that's the whole point.

**Reply by TwistedMinda** [reply]
@dai-shi Oh, okay. I get it. With the "get" keywoord i'd prevent potential eslint (or other) plugins that enforce the rules of Hooks directly inside your IDE/Text Editor to work correctly.
Very interesting, thanks.
